android/lib/src/main/java/com/swmansion/gesturehandler/TapGestureHandler.java

```diff
@@ -68,27 +88,74 @@ private void endTap() {
     }
   }
 
+  private static float getLastPointer(MotionEvent event, boolean isX) {
+    float offset = isX ? event.getRawX() - event.getX() : event.getRawY() - event.getY();
+    int excludeIndex = event.getActionMasked() == MotionEvent.ACTION_POINTER_UP ?
+            event.getActionIndex() : -1;
+      float sum = 0f;
+      int count = 0;
+      for (int i = 0, size = event.getPointerCount(); i < size; i++) {
+        if (i != excludeIndex) {
+          sum += (isX ? event.getX(i) : event.getY()) + offset;
+          count++;
+        }
+      }
+      return sum / count;
+  }
+
+  private boolean shouldFail() {
+    float dx = mLastX - mStartX + mOffsetX;
+    if (mMaxDeltaX != MAX_VALUE_IGNORE && Math.abs(dx) > mMaxDeltaX) {
+      return true;
+    }
+
+    float dy = mLastY - mStartY + mOffsetY;
+    if (mMaxDeltaY != MAX_VALUE_IGNORE && Math.abs(dy) > mMaxDeltaY) {
+      return true;
+    }
+
+    float dist = dy * dy + dx * dx;
+    return mMaxDist != MAX_VALUE_IGNORE && Math.abs(dist) > mMaxDist;
+  }
+
   @Override
   protected void onHandle(MotionEvent event) {
     int state = getState();
+    int action = event.getActionMasked();
+
+    if (action == MotionEvent.ACTION_POINTER_UP || action == MotionEvent.ACTION_POINTER_DOWN) {
+      mOffsetX += mLastX - mStartX;
+      mOffsetY += mLastY - mStartY;
+      mLastX = getLastPointer(event, true);
+      mLastY = getLastPointer(event, false);
+      mLastEventOffsetX = event.getRawX() - event.getX();
+      mLastEventOffsetY = event.getRawY() - event.getY();
+      mStartX = mLastX;
+      mStartY = mLastY;
+    } else {
+      mLastX = getLastPointer(event, true);
+      mLastY = getLastPointer(event, false);
+      mLastEventOffsetX = event.getRawX() - event.getX();
+      mLastEventOffsetY = event.getRawY() - event.getY();
+    }
 
     mLastEventOffsetX = event.getRawX() - event.getX();
     mLastEventOffsetY = event.getRawY() - event.getY();
     mLastX = event.getRawX();
     mLastY = event.getRawY();
 
-    if (state == STATE_UNDETERMINED) {
-      begin();
-      mTapsSoFar = 0;
```

Comment:
shouldn't we keep this? Have you tested multi-tap?

Reply:
Yes, please take a look on onTap method and way of invoking this method